src/ContractionTreeToGraph.jl

```diff
@@ -0,0 +1,120 @@
+        group1 = collect(Leaves(contract_sequence[1]))
+        group2 = collect(Leaves(contract_sequence[2]))
+        remaining_verts = setdiff(leaves, vcat(group1, group2))
+        add_vertex!(current_g, (group1, group2, remaining_verts))
+        spawn_child_branch(contract_sequence[1], current_g, leaves)
+        spawn_child_branch(contract_sequence[2], current_g, leaves)
+    else
+        #If it is it is just a vertex
+        add_vertex!(current_g, ([contract_sequence], setdiff(leaves, [contract_sequence])))
+    end
+end
+
+"""Utility functions for the graphical representation of a contraction sequence
+Perhaps it should be a specific object or type of tree?!"""
+
+"""Determine if a node is a leaf"""
+function leaf_node(g::AbstractGraph, v)
+    return length(neighbors(g, v)) == 1
+end
+
+"""Determine if an edge involves a leaf (at src or dst)"""
+function leaf_edge(g::AbstractGraph, e)
+    return leaf_node(g, src(e)) || leaf_node(g, dst(e))
+end 
+
+"""Determine if a node has no neighbours which are leaves"""
+function no_leaf_neighbours(g::AbstractGraph, v)
+    for vn in neighbors(g, v)
+        if(leaf_node(g, vn))
+            return false
+        end
+    end
+    return true
+end
+
+"""Get all edges which do not involve a leaf
+(why does edges(g)[contraction_edge(g, edges(g)) .== true] not work?)"""
+function non_leaf_edges(g::AbstractGraph)
+    return edges(g)[findall(==(1), [!leaf_edge(g,e) for e in edges(g)])]
+end
+
+"""Get all nodes which aren't leaves and don't have leaf neighbours (these represent internal points in the contraction sequence)"""
+function internal_nodes(g::AbstractGraph)
+    return vertices(g)[findall(==(1), [no_leaf_neighbours(g, v) && !leaf_node(g, v) for v in vertices(g)])]
+end
+
+"""Get all nodes which have leaf neighbours but aren't leaves themselves (these represent start/ end points of a contraction sequence)"""
+function external_nodes(g::AbstractGraph)
+    return vertices(g)[findall(==(1), [!no_leaf_neighbours(g, v) && !leaf_node(g,v)  for v in vertices(g)])]
+end
+
+"""Get the vertex bi-partition that a given edge between non-leaf nodes represents"""
+function edge_bipartition(g::AbstractGraph, e)
+
+    if(leaf_edge(g, e))
+        println("ERROR: EITHER THE SOURCE OR THE VERTEX IS A LEAF SO EDGE DOESN'T REALLY REPRESENT A BI-PARTITION")
+    end
+
+    vsrc_set, vdst_set = [Set(vni) for vni in src(e)], [Set(vni) for vni in dst(e)]
+    c1, c2, c3 = [src(e)[1]..., src(e)[2]...], [src(e)[2]..., src(e)[3]...], [src(e)[1]..., src(e)[3]...]
+    left_bipartition = Set(c1) ∈ vdst_set ? c1 : Set(c2) ∈ vdst_set ? c2 : c3
+
+    c1, c2, c3 = [dst(e)[1]..., dst(e)[2]...], [dst(e)[2]..., dst(e)[3]...], [dst(e)[1]..., dst(e)[3]...]
+    right_bipartition = Set(c1) ∈ vsrc_set ? c1 : Set(c2) ∈ vsrc_set ? c2 : c3
+
+    return left_bipartition, right_bipartition
+end
+
+"""Given a contraction node, get the keys living on all its neighbouring leaves"""
+function external_node_keys(g::AbstractGraph, v)
```

Comment:
It is supposed to get all the keys directly associated with the neighbouring leaves of that vertex. But I agree it is not clear/ hard to see whether it should really be a function at all.

I wrote it because when you want to figure out the tri-partition a non_leaf vertex represents one has to distinguish between keys which are coming from neighbouring leaves, and those which are not. The keys from neighbouring sources are going to be directly used in the contraction, whilst those from non-neighbouring sources just form 'environments' as opposed to being directly used.

Should I just remove for now?

Reply:
I think I understand. I think this is a good reason to first think about implementing that kind of functionality through generic `AbstractGraph` functions that work for general graphs or tree graphs, and hopefully through functions like that we can implement most functionality like this.

For example, a general graph/tree function for that might just be grabbing a subgraph in each direction away from the specified vertex and then collecting the leaves of each subgraph, which seems like a simpler function to write and easier logic to follow.

Reply:
So for now I'd say get rid of this function and we can try to come up with general `AbstractGraph` functions which help with various situations that arise in tensor network calculations when using contraction trees.

Reply:
Okay I will get rid of it and we will go that route.